pkg/metallb/frr.go

```diff
@@ -0,0 +1,417 @@
+
+	if err != nil {
+		return builder, fmt.Errorf("can not delete FRRConfiguration: %w", err)
+	}
+
+	builder.Object = nil
+
+	return builder, nil
+}
+
+// GetFrrConfigurationGVR returns FrrConfiguration GroupVersionResource which could be used for Clean function.
+func GetFrrConfigurationGVR() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group: FRRAPIGroup, Version: APIVersion, Resource: "frrconfigurations",
+	}
+}
+
+// Exists checks whether the given FrrConfiguration exists.
+func (builder *FrrConfigurationBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof(
+		"Checking if FrrConfiguration %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// WithBGPPassword defines the password used between BGP peers to form adjacency.
+func (builder *FrrConfigurationBuilder) WithBGPPassword(bgpPassword string) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define FRRConfiguration %s in namespace %s with the password: %s",
+		builder.Definition.Name, builder.Definition.Namespace, bgpPassword)
+
+	if bgpPassword == "" {
+		glog.V(100).Infof("The bgpPassword should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the bgpPassword %s is an empty string", bgpPassword)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	builder.Definition.Spec.BGP.Routers[0].Neighbors[0].Password = bgpPassword
+
+	return builder
+}
+
+// WithToReceiveModeAll allows all incoming routes.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeAll(mode string) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive all routes in namespace %s with mode type: all",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	builder.Definition.Spec.BGP.Routers[0].Neighbors[0].ToReceive.Allowed.Mode = frrtypes.AllowMode(mode)
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows all defined prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(
+	neigh frrtypes.Neighbor, prefixes []string) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"AllowMode 'filtered' receives only routes with in the prefix %s", prefixes)
+
+	for _, prefix := range prefixes {
+		glog.V(100).Infof("AllowMode 'filtered' receives only routes within the prefix %s", prefix)
+
+		if _, _, err := net.ParseCIDR(prefix); err != nil {
+			glog.V(100).Infof("The prefix of the frrConfiguration BGP peer contains an invalid IP "+
+				"address: %s", prefix)
+
+			builder.errorMsg = "frrConfiguration 'peerIP' of the BGPPeer contains an invalid IP address"
```

Comment:
looks like incorrect error message

Reply:
This was the actual error received. I will double check.

Reply:
Updated